### PR TITLE
Add single-sign-on for Mimeo to order and track offline CSF workshop materials

### PR DESCRIPTION
**Pull request by aoby**
Now, after a CSF workshop exit survey is submitted, this will be displayed:

image
---
Clicking the link goes here briefly:

image
---
Then automatically redirects to Mimeo with the CSF workshop materials in the cart, and only one item allowed:

image

### CODE REVIEW

File: dashboard/app/mailers/pd/workshop_mailer.rb [hunk 1]
```diff
@@ -153,12 +153,7 @@ def exit_survey(enrollment)
     @workshop = enrollment.workshop
     @teacher = enrollment.user
     @enrollment = enrollment
-
-    if [Pd::Workshop::COURSE_ADMIN, Pd::Workshop::COURSE_COUNSELOR].include? @workshop.course
-      @survey_url = CDO.code_org_url "/pd-workshop-survey/counselor-admin/#{enrollment.code}", 'https:'
-    else
-      @survey_url = CDO.code_org_url "/pd-workshop-survey/#{enrollment.code}", 'https:'
-    end
+    @survey_url = enrollment.exit_survey_url
```
Comment (aoby) [inline dashboard/app/mailers/pd/workshop_mailer.rb:156]: This refactor is not directly related, but since I was messing with the exit survey I moved this definition into `Pd::Enrollment` and added a unit test.

File: pegasus/sites.v3/code.org/public/pd-workshop-survey/splat.haml [hunk 2]
```diff
@@ -305,3 +304,13 @@ title: "Professional Development Workshop Survey"
   #thanks{style: 'display: none;'}
     %h1 Thank you for your submission
 
+- if mimeo_url
+  #materials-order{style: 'display: none;'}
+    %p
+      Click
```
Comment (mehalshah) [inline pegasus/sites.v3/code.org/public/pd-workshop-survey/splat.haml:310]: We should say a little about what mimeo is so people don't think we are just redirecting them to some random website

File: dashboard/app/controllers/pd/mimeo_sso_controller.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      # Check pegasus forms directly before giving up.
+      raise CanCan::AccessDenied unless PEGASUS_DB[:forms].where(kind: 'PdWorkshopSurvey', source_id: @enrollment.id).any?
+    end
+
+    # Set up fields for the Mimeo SSO API. Each of these will be rendered as a hidden field on a form which will be
+    # automatically posted to the @endpoint on page load.
+    # Mimeo's API will then authenticate and either redirect the user's browser to the redirectUrl on success,
+    # or to our errorUrl if it fails (see #error below).
+    @endpoint = ENDPOINT
+    @fields = {
```
Comment (mehalshah) [inline dashboard/app/controllers/pd/mimeo_sso_controller.rb:32]: Any risk of these being visible? Users can pause the load and see these secrets

File: dashboard/app/models/pd/enrollment.rb [hunk 4]
```diff
@@ -97,6 +101,14 @@ def in_section?
     workshop.section.students.exists?(user.id)
   end
 
+  def exit_survey_url
```
Comment (mehalshah) [inline dashboard/app/models/pd/enrollment.rb:104]: Should this be a part of this change?

File: dashboard/app/views/pd/mimeo_sso/error.haml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+%p
+  = @message
+
+%p
+  Error Code:
+  = @code
+
+%p
+= link_to 'try again', action: 'authenticate_and_redirect', enrollment_code: params[:enrollment_code]
```
Comment (mehalshah) [inline dashboard/app/views/pd/mimeo_sso/error.haml]: Should we have a link to zendesk here?

File: dashboard/test/controllers/pd/mimeo_sso_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    assert_response :success
+  end
+
+  test 'logged in users cannot access enrollments they dont own' do
+    sign_in create(:teacher)
+    get :authenticate_and_redirect, params: {enrollment_code: @enrollment.code}
+    assert_response :forbidden
+  end
+
+  test 'not logged in users are redirected to login page' do
```
Comment (mehalshah) [inline dashboard/test/controllers/pd/mimeo_sso_controller_test.rb:43]: Add a test that students cannot go here?

File: dashboard/app/controllers/pd/mimeo_sso_controller.rb [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      firstName: encrypt_token(@enrollment.first_name),
+      lastName: encrypt_token(@enrollment.last_name),
+      email: encrypt_token(current_user.email),
```
Comment (mehalshah) [inline dashboard/app/controllers/pd/mimeo_sso_controller.rb]: Are we guaranteed to have all these fields? It's possible for last_name to be blank if this is old (and arguably we shouldn't be requiring last name)

File: dashboard/app/controllers/pd/mimeo_sso_controller.rb (hunk 3 above)
Comment (aoby) [reply dashboard/app/controllers/pd/mimeo_sso_controller.rb:32]: In Mimeo's [example C# code](https://github.com/Mimeo/Connect/blob/master/SSO/RSA%20Examples/Mimeo-SSO.CS), instead of javascript auto-submitting, there is a button the user has to press in order to submit the form and start the process. It's designed such that these form fields may be visible. I chose to auto-submit to remove the extra step of pushing a button. I also saw examples of doing it this way online. Pausing is not really a big deal. 

Most of the fields are encrypted with Mimeo's RSA public key, which is not exposed. Only Mimeo has the private key to decrypt. The encrypted fields are `firstName`, `lastName`, `email`, `organizationId`, and `companyName`. Strangely their API expects `companyId` to be unencrypted.

So it's true a savvy user could scrape and reuse the fields, including the encrypted ones. However if they reuse their own name and email, they will simply log back into the same account. And in order to spoof and create a new account, they will need to have the RSA public key (which is secret) to encrypt a new name and email that is valid and can be decrypted and understood on the other end.

File: dashboard/app/controllers/pd/mimeo_sso_controller.rb (hunk 7 above)
Comment (aoby) [reply dashboard/app/controllers/pd/mimeo_sso_controller.rb]: We do [require](https://github.com/code-dot-org/code-dot-org/blob/a2adb2db73b4a17d6cbc5d4cfd9145eae13ba8e1/dashboard/app/models/pd/enrollment.rb#L37-L41) first and last names for new workshop enrollments. It is true that some old enrollments might have no last name. These should have already taken the survey, but anyway it's possible. 

I just verified that a nil or empty last name causes an error:
image

So I'll go ahead and substitute a placeholder `-` in the case of blank last name.

File: dashboard/app/models/pd/enrollment.rb (hunk 4 above)
Comment (aoby) [reply dashboard/app/models/pd/enrollment.rb:104]: As I said above, it's not directly related. Initially it was because I was thinking about redirecting back to the survey in case it wasn't filled out, but I decided against it. I still think it's an improvement, but I'm happy to move it into its own PR if you prefer.

File: dashboard/app/views/pd/mimeo_sso/error.haml (hunk 5 above)
Comment (aoby) [reply dashboard/app/views/pd/mimeo_sso/error.haml]: Good call. Will add.

**Comment by aoby** [comment]
@mehalshah Updated. PTAL
